src/core/manager.c

```diff
@@ -1715,6 +1719,11 @@ static int manager_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t
         }
 
         n = recvmsg(m->notify_fd, &msghdr, MSG_DONTWAIT|MSG_CMSG_CLOEXEC);
+        if (n == 0) {
+            log_warning("Got zero-length notification message. Ignoring.");
+            return 0;
+        }
+
```

Comment:
The (n == 0) check should be *after* the (n < 0). We check for error conditions first, and then bad input second.

Also, please don't add an empty line here. Since those are all checks for the same operation, they should be adjacent.

Finally, we don't want to warn here, for a repeateable operation performed by some broken service. A debug level log would be OK.